nx/lib/nx/shape.ex

```diff
@@ -556,6 +556,78 @@ defmodule Nx.Shape do
     end
   end
 
+  @doc """
+  Flattens the given axes of the given input shape into a
+  single axis.
+
+  ## Examples
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [nil, nil, nil], nil)
+      {{6}, [nil]}
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [:batch, nil, nil], [1, 2])
+      {{1, 6}, [:batch, nil]}
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [nil, nil, nil], [])
+      {{1, 2, 3}, [nil, nil, nil]}
+
+  ### Error cases
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [:batch, nil, nil], [0, 2])
+      ** (ArgumentError) flatten axes must be consecutive
+  """
+  def flatten(shape, names, axes)
+
+  def flatten({}, _, _), do: {{}, []}
+
+  def flatten(shape, names, []), do: {shape, names}
+
+  def flatten(shape, names, axes) do
+    axes = axes || Enum.to_list(0..(tuple_size(shape) - 1))
+    [insert_axis | _] = axes = normalize_axes(shape, axes, names)
+
+    unless consecutive?(axes) do
+      raise ArgumentError, "flatten axes must be consecutive"
+    end
+
+    shape = Tuple.to_list(shape)
+    shape_and_names = Enum.zip(shape, names)
+
+    {rev_shape, rev_names, flatten_size} =
+      for {{size, name}, i} <- Enum.with_index(shape_and_names), reduce: {[], [], 1} do
+        {cur_shape, cur_names, cur_size} ->
+          if i in axes do
+            {cur_shape, cur_names, size * cur_size}
+          else
+            {[size | cur_shape], [name | cur_names], cur_size}
+          end
+      end
```

Comment:
Nitpick: since we already have a loop, might as well "fuse" the `with_index` operation:

```suggestion
    {rev_shape, rev_names, flatten_size, _} =
      for {size, name} <- shape_and_names, reduce: {[], [], 1, 0} do
        {cur_shape, cur_names, cur_size, i} ->
          if i in axes do
            {cur_shape, cur_names, size * cur_size, i + 1}
          else
            {[size | cur_shape], [name | cur_names], cur_size, i + 1}
          end
      end
```